pkg/segment/writer/segstore.go

```diff
@@ -220,7 +221,174 @@ func (segstore *SegStore) resetSegStore(streamid string, virtualTableName string
+				addIntToRangeIndex(colName, intVal, rangeIndex)
+				continue
+			}
+
+			// Try converting to a float.
+			floatVal, err := strconv.ParseFloat(numberAsString, 64)
+			if err == nil {
+				// Conversion succeeded.
+				copy(newColWip.cbuf[newColWip.cbufidx:], utils.VALTYPE_ENC_FLOAT64[:])
+				copy(newColWip.cbuf[newColWip.cbufidx+1:], toputils.Float64ToBytesLittleEndian(floatVal))
+				newColWip.cbufidx += 1 + 8
+				addFloatToRangeIndex(colName, floatVal, rangeIndex)
+				continue
+			}
+
+			// Conversion failed.
+			return false
+
+		case utils.VALTYPE_ENC_INT64[0], utils.VALTYPE_ENC_FLOAT64[0]:
+			// Already a number, so just copy it.
+			// It's alrady in the range index, so we don't need to add it again.
+			copy(newColWip.cbuf[newColWip.cbufidx:], oldColWip.cbuf[i-1:i+8])
+			newColWip.cbufidx += 9
+			i += 8
+
+		case utils.VALTYPE_ENC_BACKFILL[0]:
+			// This is a null value.
+			copy(newColWip.cbuf[newColWip.cbufidx:], utils.VALTYPE_ENC_INT64[:])
+			copy(newColWip.cbuf[newColWip.cbufidx+1:], toputils.Int64ToBytesLittleEndian(0))
+			newColWip.cbufidx += 1 + 8
+			addIntToRangeIndex(colName, 0, rangeIndex)
+
+		case utils.VALTYPE_ENC_BOOL[0]:
+			// Cannot convert bool to number.
+			return false
+
+		default:
+			// Unknown type.
+			log.Errorf("convertColumnToNumbers: unknown type %v", valType)
+			return false
+		}
+	}
+
+	// Conversion succeeded, so replace the column with the new one.
+	wipBlock.colWips[colName] = newColWip
+	delete(wipBlock.columnBlooms, colName)
+	return true
+}
+
+func convertColumnToStrings(wipBlock *WipBlock, colName string, segmentKey string) {
+	oldColWip := wipBlock.colWips[colName]
+	newColWip := InitColWip(segmentKey, colName)
+	bloom := wipBlock.columnBlooms[colName]
+
+	for i := uint32(0); i < oldColWip.cbufidx; {
+		valType := oldColWip.cbuf[i]
+		i++
+
+		switch valType {
+		case utils.VALTYPE_ENC_SMALL_STRING[0]:
+			// Already a string, so just copy it.
+			// This is already in the bloom, so we don't need to add it again.
+			numBytes := uint32(toputils.BytesToUint16LittleEndian(oldColWip.cbuf[i : i+2]))
+			i += 2
+			copy(newColWip.cbuf[newColWip.cbufidx:], oldColWip.cbuf[i-3:i+numBytes])
+			newColWip.cbufidx += 3 + numBytes
+			i += numBytes
+
+		case utils.VALTYPE_ENC_INT64[0]:
+			// Parse the integer.
+			intVal := toputils.BytesToInt64LittleEndian(oldColWip.cbuf[i : i+8])
+			i += 8
+
+			stringVal := strconv.FormatInt(intVal, 10)
+			newColWip.WriteSingleString(stringVal)
+			bloom.uniqueWordCount += addToBlockBloom(bloom.Bf, []byte(stringVal))
+
+		case utils.VALTYPE_ENC_FLOAT64[0]:
+			// Parse the float.
+			floatVal := toputils.BytesToFloat64LittleEndian(oldColWip.cbuf[i : i+8])
+			i += 8
+
+			stringVal := strconv.FormatFloat(floatVal, 'f', -1, 64)
+			newColWip.WriteSingleString(stringVal)
+			bloom.uniqueWordCount += addToBlockBloom(bloom.Bf, []byte(stringVal))
+
+		case utils.VALTYPE_ENC_BACKFILL[0]:
+			// This is a null value.
+			newColWip.WriteSingleString("")
```

Comment:
for backfill, we don't write any actual value ? we just store the valType, so we just need to copy the VALTYPE_...BACKFILL (just that 1 byte)